src/Hamiltonians/HOCartesianEnergyConserved.jl

```diff
@@ -0,0 +1,376 @@
+* `S`: Tuple of the number of levels in each dimension, including the groundstate. The 
+    allowed couplings between states is defined by the aspect ratio of `S .- 1`. Defaults 
+    to a 1D spectrum with number of levels matching modes of `addr`. Will be sorted to make 
+    the first dimension the largest.
+* `η`: Define a custom aspect ratio for the trapping potential strengths, instead of deriving
+    from `S .- 1`. This will only affect the single particle energy scale and not the 
+    interactions. The values are always scaled relative to the first dimension, which sets 
+    the energy scale of the system, ``\\hbar\\omega_x``.
+* `g`: the (isotropic) interparticle interaction parameter. The value of `g` is assumed 
+    to be in trap units.
+* `interaction_only`: if set to `true` then the noninteracting single-particle terms are 
+    ignored. Useful if only energy shifts due to interactions are required.
+
+See also [`HOCartesianEnergyConservedPerDim`](@ref).
+
+!!! warning
+    `num_offdiagonals` is a bad estimate for this Hamiltonian. Take care when building 
+    a matrix or using QMC methods. Use [`get_all_blocks`](@ref) first then pass option
+    `col_hint = block_size` to [`BasisSetRep`](@ref) to safely build the matrix.
+"""
+struct HOCartesianEnergyConserved{
+    D,  # number of dimensions
+    A<:BoseFS,
+    T
+} <: AbstractHamiltonian{Float64}
+    addr::A
+    S::NTuple{D,Int64}
+    aspect::NTuple{D,T}
+    aspect1::NTuple{D,Float64}  # (possibly custom) ratios of trapping frequencies for single-particle energy scale
+    energies::Vector{Float64}   # noninteracting single particle energies
+    vtable::Array{Float64,4}    # interaction coefficients
+    u::Float64
+end
+
+function HOCartesianEnergyConserved(
+        addr::BoseFS; 
+        S = (num_modes(addr),),
+        η = nothing, 
+        g = 1.0,
+        interaction_only = false
+    )
+    # x dimension defines the energy scale, and should be the smallest frequency i.e. largest spatial dimension
+    D = length(S)
+    S_sort = tuple(sort(collect(S); rev = true)...)
+    S == S_sort || @warn("dimensions have been reordered")
+    
+    P = prod(S)
+    P == num_modes(addr) || throw(ArgumentError("number of modes does not match starting address"))
+
+    aspect = box_to_aspect(S)
+    if all(denominator.(aspect) .== 1)
+        aspect = Int.(aspect)
+    end
+
+    if isnothing(η)
+        aspect1 = float.(aspect)
+    elseif length(η) == D
+        aspect1 = S == S_sort ? η : η[sortperm(collect(S); rev = true)]
+    elseif length(η) == 1 && isreal(η) && η ≥ 1
+        aspect1 = ntuple(d -> (d == 1 ? 1.0 : η), D)
+    else
+        throw(ArgumentError("Invalid aspect ratio parameter η."))
+    end
+
+    if interaction_only
+        energies = zeros(P)
+    else
+        states = CartesianIndices(S)    # 1-indexed
+        energies = reshape(map(x -> dot(aspect1, Tuple(x) .- 1/2), states), P)
+    end
+
+    u = sqrt(prod(aspect1)) * g / 2
```

Comment:
Why is the interaction energy scaled like this?

Reply:
It comes from a change of variables when evaluating the integrals for the interaction matrix elements. But it should actually be divided by `sqrt(prod(aspect1))`